### ISSUE TITLE
KeeWeb Security: Is it still safe to use Keeweb in 2023?

### ISSUE THREAD

**Issue by howdytom** [keeweb/keeweb#2077]
Happy New Year! 

Since there has been no development and the project is on hold and probably EOL - What about the KeeWebs security? Is it still safe to use KeeWeb for Mac/Windows in 2023? 

Any replies are highly appreciated.

**Comment by jivarson** [comment]
Id say the core is safe, but the dependencies could be vulnerable. If there's a risk it would be a sideloading-attack or such on the clientside. If you're worried, you should switch to KeePassXC which i compatible. You have to sync the database yourself though.

I think theres bigger risk that stuff like signing-certs or syncronization and stuf fthat requires active maintance will stop work first.. or OS-compat to be outdated.

**Comment by howdytom** [comment]
I absolutely agree and I was expecting this kind of answer. This is exactly why I use Keewebs core only, without any dependencies which is already inconvenient. Thank you. I really appreciate your reply.

**Comment by strazto** [comment]
Though the dependencies **could** be vulnerable, you do have the option of forking it and then running dependabot against your fork, merging whatever dependabot flags, then building from there.

Overall, that should be a reasonably low effort way to mitigate supply chain vulns.

A fork that just comprises SCA automations & releases would be amazing to see.

I do think it's unfortunate that the project is no longer maintained, and #2022 hasn't seemed to gain any apparent traction, whether that's due to a lack of applicants, or a lack of quality applicants, or a lack of time to actually handover to the new maintainer.

Related: 

Can anyone suggest a good Windows client alternative?
I use Keepass2Android on my phone, and it's excellent.

On MacOS I would probably turn to Strongbox (I don't mind their pricing model), same for iPhone if I had an iPhone.

The only gap is Windows - I'd be happy enough to keep using keeweb, (especially via a fork that includes SCA mitigations) but I wonder if there's a maintained alternative that has a more "native" UX.

**Comment by strazto** [comment]
In fact, I've gone ahead and soft-forked this & enabled dependabot, I'll respond to merge requests by dependabot and try to publish releases whenever I do that.

https://github.com/strazto/keeweb

**Comment by thefiredragon** [comment]
@strazto will you also do some releases next time ?

**Comment by strazto** [comment]
Oh lol dependable is actually annoying as hell so I got distracted, I'll have another crack

**Comment by HarlemSquirrel** [comment]
I offered to help maintain like a year ago or more but got no response. I'm not in the EU so maybe that's why.

**Comment by HarlemSquirrel** [comment]
@strazto I just found this which looks promissing: https://github.com/OneKeePass/desktop

**Comment by Aetherinox** [comment]
v1.19 will fix any vulnerabilities that existed in older packages that are currently in-use. The beta version right now is using mostly updated packages, aside from Electron, which got a slight bump, but we need to re-organize the framework in order to update Electron any further.

As long as your system itself isn't compromised; you should have no issues maintaining a secure KeepWeb database.
- https://github.com/keeweb/keeweb/pull/2168

**Comment by strazto** [comment]
Good to see the project has picked up new maintainership.

Thank you for addressing this @Aetherinox

**Comment by KaKi87** [comment]
Well, looks like it's dying again

**Comment by Aetherinox** [comment]
Nope, not dead. I have a Gitea repo that I develop on, and only push to Github once a finalized function is complete. I am on this repo almost every day. 

Coupled with the fact that I had the flu recently which put me about two weeks behind, and the biggest task at the moment is working with Google so that we can get our Google Drive full scope back as discussed in another thread.

But as I've said all-along, if others have experience (fluent) with Javascript / node / electron, and wish to help, you guys are more than welcome to submit PRs, as long as the code is proper. 

For the time being, the beta website has been updated, and there's a much more recent docker image of the next version of KeeWeb. The only version yet to be updated is the official binaries.

And also another library utilized by KeeWeb had to get a massive overhaul since it has not been updated in quite some time, which introduced a potential security vulnerability into the app.

Ant has been busy with personal life, and I'm not going expect him to help. He trusted in the fact that this app would be managed properly. But as I've said, if others want quicker progression, and know the code, by all means, I'm not stopping others from helping.

**Comment by KaKi87** [comment]
> I have a Gitea repo that I develop on

Oh, what's the link ? Also why not Forgejo ?

> I had the flu recently

I'm glad it's over for you !

> the biggest task at the moment is working with Google so that we can get our Google Drive full scope back

Honestly, I'd vote in favor of dropping Google Drive and keeping only those with reasonable expectations and communication from developers. Between this and Manifest V3, Google really sucks.

> if others have experience (fluent) with Javascript / node / electron, and wish to help, you guys are more than welcome to submit PRs

What front-end framework is used on the project ? I tried to figure that out but I couldn't.

> The only version yet to be updated is the official binaries.

And the HTML archive for static file hosting.

Thank you

**Comment by Aetherinox** [comment]
> Oh, what's the link ?

Localhost copy. Just in case there's any accidents (like secrets), I don't expose them. Plus I have secret scanning enabled, so I can be informed in case an accident happens. Plus I like to run my own runners. Github gives you 2,000 free minutes, but I run a lot of runners in a day, so I could easily eat that up in probably 2 weeks.

> Also why not Forgejo ?

Honestly, never tried it. Gitea was the first program I used when I started my "self-hosting" adventure. And I've just liked it. Workflows work great, no issue with runners, and I've even got my Gitea theme pretty close to Github's layout. Gitea's layout was alright, but I'm used to Github. So I spent some time "re-arranging" how things are formatted. Plus Gitea has worked great for Docker images (other than the fact that Portainer doesn't support image browsing for a Gitea registry.

I've heard about Forgejo, GOGS, and a few others, but I just never tried them out. I don't remember which one, but I think I heard that Gitea and one other used to be the same company, and then they branched off with some old developers. Don't remember which one it was though.

> Honestly, I'd vote in favor of dropping Google Drive and keeping only those with reasonable expectations and communication from developers. Between this and Manifest V3, Google really sucks.

Removing Google drive would be easier, but it would kill something that many users rely on. The only thing that makes this process a pain is that Google has upped their requirements / application process if you happen to be a company in need of the scope we require, which allows you to browse your entire Google Drive, instead of just the root folder.

But, while I've been going through this fiasco, there may be another alternative. The downside is that it's a 3rd party library, so if the developer ever went MIA one day, we'd be on our own. The up side is that it does not require us to get approval from Google. Google has nothing to do with the library. So there's benefits and negative affects.

> What front-end framework is used on the project ? I tried to figure that out but I couldn't.

Not sure which part you're referring to, mainly because things vary between the different builds. In short, Node for the overall package management / structure, Electron for the binaries. Electron is the wrapper which allows KeeWeb to store data similar to how your browser can, through a local DB, cached store, etc. 

Webpack takes care of the local server requirements, so that you can run a local webserver to host KeeWeb, without having to expose it publically, and don't want to use the local html files.

Grunt for task management, the process of storing passwords is handled by kdbxweb (an in-house javascript library). 

But when you get down to the front-end code, it's nothing more than HTML, CSS, and Javascript. That's what makes KeeWeb light-weight, and can be ran from a single HTML file which is compressed down into a 2500kb + file. 

The binaries on the other side have to have Electron packaged with them, since Electron is what will serve as the main framework for Keeweb to run within.

> And the HTML archive for static file hosting.

I'll have to check, I don't remember if I have or haven't updated the static files. I swear I did, but then again, I've had a lot of different parts I've been working on, so I may have missed it. Or maybe I'm thinking about the beta keeweb repo, and I forgot to update the static files. I'll have to check.

**Comment by KaKi87** [comment]
> Github gives you 2,000 free minutes, but I run a lot of runners in a day, so I could easily eat that up in probably 2 weeks.

You might be interested in Codeberg, it's a non-profit running Forgejo with CI/CD fully free for everyone.

> my Gitea theme pretty close to Github's layout

Oh, which one ?

> I've heard about Forgejo, GOGS, and a few others, but I just never tried them out. I don't remember which one, but I think I heard that Gitea and one other used to be the same company, and then they branched off with some old developers. Don't remember which one it was though.

Gogs is the original project, Gitea is a fork created by the community because of a disagreement by a majority of contributors with Gogs' maintainer, then Forgego is a fork of Gitea created again by the community because maintainers decided to put Gitea under the direction of a company with trademark registration and later to make it open core.

> it would kill something that many users rely on

I would hope users themselves would want to run away from Google if they know how they're treating their favourite KDBX manager's developers.

> a 3rd party library [...] does not require us to get approval from Google

How come ?

---

PS : this comment is incomplete, I misclicked the submit button and will continue writing it, sorry

**Comment by KaKi87** [comment]
> when you get down to the front-end code, it's nothing more than HTML, CSS, and Javascript

Oh ! Does it follow a structure or a pattern ? Is it documented somewhere ?

Honestly, vanilla is the most difficult to encourage contributions.

> That's what makes KeeWeb light-weight, and can be ran from a single HTML file which is compressed down into a 2500kb + file.

Even apps running a frontend framework can achieve this, really.

> I don't remember if I have or haven't updated the static files

Those are uploaded as ZIP files in GitHub releases, the last one is from 2021 : [`KeeWeb-1.18.7.html.zip`](https://github.com/keeweb/keeweb/releases/download/v1.18.7/KeeWeb-1.18.7.html.zip)

**Comment by Aetherinox** [comment]
> Oh ! Does it follow a structure or a pattern ?

It's pretty vanilla in terms of the front end. Obviously on the backend, it's not so vanilla. We utilize SCSS with a particular library, and we have to keep up with that library as far as deprecations. So if anyone wanted to develop a new feature for KeeWeb, they have to conform their code with that library. But there's plenty of documentation.

And then just the standard rules, we utilize content policies, which disallows fetching images, css, js, fonts, etc. from external websites and allowing external websites from loading it into KeeWeb Obviously for security reasons. So if any new images or Javascript are to be added, they must be local to the app. They can't just be called from an external URL otherwise they'll be blocked from loading.

> Is it documented somewhere ?

Sadly, documentation is lacking. When I generated our new docker images, I made sure to document them in..... probably too much detail. But I did that with the purpose that if anyone wanted to do anything with the image, they had a blueprint on how to do it. Obviously Ant can't be blamed. KeeWeb development is a lot of work, and adding detailed documentation eats up even more time.

I have plans in mind that I want to migrate us to Mkdocs, and host it on a Github pages workflow, and I'll write the documentation as I go. I just haven't had time to work on that project yet because many other things take priority.

> Even apps running a frontend framework can achieve this, really.

True. Ant did a good job at ensuring that KeeWeb had multiple deployment options. And he did a good job of ensuring everything was compressed to a point where you weren't carrying around bloat. When I use these types of apps, I much prefer having a light portable version. The electron version is nice, but it's better having a permanent home. 

> Those are uploaded as ZIP files in GitHub releases, the last one is from 2021 : [KeeWeb-1.18.7.html.zip](https://github.com/keeweb/keeweb/releases/download/v1.18.7/KeeWeb-1.18.7.html.zip)

Ah. I see what you mean now. As I dig around, I released new static files for the repo, but I never pushed out an official release yet, and that was mainly because of us waiting on the binaries to get sorted. I just finished updating the distribution workflows the day before yesterday, which I'm testing now.

**Comment by KaKi87** [comment]
> I want to migrate us to Mkdocs, and host it on a Github pages workflow, and I'll write the documentation as I go. I just haven't had time to work on that project yet because many other things take priority.

Would you consider increasing the priority of that task in order to make external contributions easier ?

For example, I would like to know how to read and improve components, and how to make new ones, so that I could work on #1458, #1459, #1837 and #1576.

> I never pushed out an official release yet

I'm looking forward to it !

**Comment by KaKi87** [comment]
Hi,
Any news on an official release please ?
Thanks

**Comment by Aetherinox** [comment]
Check the PR, I posted a message in there a day or two ago.

**Comment by Harald-Koschinski** [comment]
@Aetherinox can you please post a link to your message

**Comment by KaKi87** [comment]
https://github.com/keeweb/keeweb/issues/2160#issuecomment-2843255501

**Comment by strazto** [comment]
> > the biggest task at the moment is working with Google so that we can get our Google Drive full scope back
> 
> Honestly, I'd vote in favor of dropping Google Drive and keeping only those with reasonable expectations and communication from developers

Personally I'd love to see native Google drive come back, and if you're able to get traction with Google to that end I'll be very glad.

In the meantime, for anyone who needs to figure out how they can use a GDrive backed database, my suggestion is to use the Google drive desktop app to mount a `G://` drive, and then point the KeeWeb app at that file path instead.

I can't speak to how this works on Linux or Mac - I don't use desktop Linux anymore (only headless), and on Mac I exclusively use Strongbox to access the same db, which I paid for with no regrets.
